src/cluster.c

```diff
@@ -5333,48 +5333,58 @@ void addNodeToNodeReply(client *c, clusterNode *node) {
     } else {
         serverPanic("Unrecognized preferred endpoint type");
     }
-    
+
     /* Report non-TLS ports to non-TLS client in TLS cluster if available. */
     int use_pport = (server.tls_cluster &&
                      c->conn && (c->conn->type != connectionTypeTls()));
     addReplyLongLong(c, use_pport && node->pport ? node->pport : node->port);
     addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
 
     /* Add the additional endpoint information, this is all the known networking information
-     * that is not the preferred endpoint. */
-    void *deflen = addReplyDeferredLen(c);
+     * that is not the preferred endpoint. Note the logic is evaluated twice so we can
+     * correctly report the number of additional network arguments without using a deferred
+     * map. */
     int length = 0;
+    if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_IP) {
+        length++;
+    }
+    if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_HOSTNAME
+        && sdslen(node->hostname) != 0)
+    {
+        length++;
+    }
+    addReplyMapLen(c, length);
+
     if (server.cluster_preferred_endpoint_type != CLUSTER_ENDPOINT_TYPE_IP) {
         addReplyBulkCString(c, "ip");
         addReplyBulkCString(c, node->ip);
-        length++;
```

Comment:
When we have the same logic in two places, it's potentially risky. Would it make sense to assert that they're matching in some way? E.g. decrement length here and assert that it's zero in the end?

Reply:
I did think about it, but it seemed so verbose at the time. It's probably the right thing though.

Reply:
OK, you decide if it's worth it or if it's too verbose. =)

Reply:
I decided it was worth it :)